RecoTracker/LSTCore/src/alpaka/Event.dev.cc

```diff
@@ -0,0 +1,1668 @@
+    segments += it;
+  }
+  for (auto& it : n_segments_by_layer_endcap_) {
+    segments += it;
+  }
+
+  return segments;
+}
+
+unsigned int Event::getNumberOfSegmentsByLayer(unsigned int layer) {
+  if (layer == 6)
+    return n_segments_by_layer_barrel_[layer];
+  else
+    return n_segments_by_layer_barrel_[layer] + n_segments_by_layer_endcap_[layer];
+}
+
+unsigned int Event::getNumberOfSegmentsByLayerBarrel(unsigned int layer) { return n_segments_by_layer_barrel_[layer]; }
+
+unsigned int Event::getNumberOfSegmentsByLayerEndcap(unsigned int layer) { return n_segments_by_layer_endcap_[layer]; }
+
+unsigned int Event::getNumberOfTriplets() {
+  unsigned int triplets = 0;
+  for (auto& it : n_triplets_by_layer_barrel_) {
+    triplets += it;
+  }
+  for (auto& it : n_triplets_by_layer_endcap_) {
+    triplets += it;
+  }
+
+  return triplets;
+}
+
+unsigned int Event::getNumberOfTripletsByLayer(unsigned int layer) {
+  if (layer == 6)
+    return n_triplets_by_layer_barrel_[layer];
+  else
+    return n_triplets_by_layer_barrel_[layer] + n_triplets_by_layer_endcap_[layer];
+}
+
+unsigned int Event::getNumberOfTripletsByLayerBarrel(unsigned int layer) { return n_triplets_by_layer_barrel_[layer]; }
+
+unsigned int Event::getNumberOfTripletsByLayerEndcap(unsigned int layer) { return n_triplets_by_layer_endcap_[layer]; }
+
+int Event::getNumberOfPixelTriplets() {
+  auto nPixelTriplets_buf_h = cms::alpakatools::make_host_buffer<unsigned int[]>(queue, 1u);
+
+  alpaka::memcpy(queue, nPixelTriplets_buf_h, pixelTripletsBuffers->nPixelTriplets_buf);
+
+  return *nPixelTriplets_buf_h.data();
+}
+
+int Event::getNumberOfPixelQuintuplets() {
+  auto nPixelQuintuplets_buf_h = cms::alpakatools::make_host_buffer<unsigned int[]>(queue, 1u);
+
+  alpaka::memcpy(queue, nPixelQuintuplets_buf_h, pixelQuintupletsBuffers->nPixelQuintuplets_buf);
+
+  return *nPixelQuintuplets_buf_h.data();
+}
+
+unsigned int Event::getNumberOfQuintuplets() {
+  unsigned int quintuplets = 0;
+  for (auto& it : n_quintuplets_by_layer_barrel_) {
+    quintuplets += it;
+  }
+  for (auto& it : n_quintuplets_by_layer_endcap_) {
+    quintuplets += it;
+  }
+
+  return quintuplets;
+}
+
+unsigned int Event::getNumberOfQuintupletsByLayer(unsigned int layer) {
+  if (layer == 6)
+    return n_quintuplets_by_layer_barrel_[layer];
+  else
+    return n_quintuplets_by_layer_barrel_[layer] + n_quintuplets_by_layer_endcap_[layer];
+}
+
+unsigned int Event::getNumberOfQuintupletsByLayerBarrel(unsigned int layer) {
+  return n_quintuplets_by_layer_barrel_[layer];
+}
+
+unsigned int Event::getNumberOfQuintupletsByLayerEndcap(unsigned int layer) {
+  return n_quintuplets_by_layer_endcap_[layer];
+}
+
+int Event::getNumberOfTrackCandidates() {
+  auto nTrackCandidates_buf_h = cms::alpakatools::make_host_buffer<unsigned int[]>(queue, 1u);
+
+  alpaka::memcpy(queue, nTrackCandidates_buf_h, trackCandidatesBuffers->nTrackCandidates_buf);
+
+  return *nTrackCandidates_buf_h.data();
```

Comment:
@makortel 
I thought that the host buffer destructor would synchronize in this case around return. However, without `alpaka::wait` the results are wrong.
Is it expected to have a `wait` here?

Reply:
The `data()` is accessed before the destruction of the buffer, so this case will need a synchronization. 

Furthermore, in this case `nTrackCandidates_buf_h` is created via `make_host_buffer<...>(queue, ...)` and thus by default he underlying allocation goes via the caching allocator, in which case the buffer destructor would not synchronize anyway.

By the way, you could use `make_host_buffer<unsigned int>(queue)` for a scalar buffer instead of 1-element array.

Reply:
> By the way, you could use `make_host_buffer<unsigned int>(queue)` for a scalar buffer instead of 1-element array.

this would need the size specified in the copy. So, the number of characters probably the same.

Reply:
what's the cost of the extra `alpaka::wait`?
The cost of figuring out how to do it right (not calling it "just to be safe") so far seems larger.

Reply:
> > By the way, you could use `make_host_buffer<unsigned int>(queue)` for a scalar buffer instead of 1-element array.
> 
> this would need the size specified in the copy. So, the number of characters is probably the same.

No, a scalar alpaka buffer knows its size, just line an N-dimensional buffer.

Reply:
> The cost of figuring out how to do it right (not calling it "just to be safe") so far seems larger.

There is no guarantee of any implicit synchronisation in the buffers, no matter if they are allocated with a `queue` or a `device` argument.

Some backends my _cause_ an extra synchronisation, but AFAIK not all of them do, and alpaka does not provide any such guarantee.

AFAIK, the behaviour that alpaka strives to implement is:
   - all operations along a `queue` are guaranteed to happen in order, and the result of an operation is visible to the next one along the same `queue`;
   - all operations along a non-blocking `queue` are potentially asynchronous with respect to the host.

So the requirement is that, after a `memcpy`, the results are visible to all following operations along the same `queue`, but are not visible to the host without an explicit synchronisation.

Bottom line, after a `memcpy` form device to host, one needs to `wait` before the results can be accessed.

Reply:
> what's the cost of the extra `alpaka::wait`?

Depends on
* the probability for the `queue` to be busy at the time of the `alpaka::memcpy()` call, and if it is busy, how long will it take for all the preceding work to finish
* the probability for concurrent device-to-host copies (at least on NVIDIA GPUs those are limited to one at a time)
* frequency of other concurrent Alpaka API calls that may on some backends use locks in the vendor runtime/driver library (NVIDIA does, but I don't remember how bad `cudaStreamSynchronize()`, that `alpaka::wait()` calls in this case, is in this regard. Or maybe I've never seen it in the threading profiles because we haven't used it for synchronization)

Then of course there is a backend-dependent constant cost with the call ("overhead"), but I consider (guess) that to be negligible compared to the other points above.

If LST is run alone, I can believe 

> The cost of figuring out how to do it right (not calling it "just to be safe") so far seems larger.

I can believe "doing it right" would require larger reorganization. Since LST is not targeting production any time soon, I think leaving the `wait()` here to be addressed later (along all the other `wait()` calls) is reasonable. To me the scope of this PR for not-easily-removable `wait()` calls is to understand (and document) their need.

Reply:
at least in this specific case, it sounds like an explicit `wait` is unavoidable.

Reply:
As long as the `getNumberOfTrackCandidates()` is used only in debug prints (that are disabled by default), I think the explicit `wait()` is acceptable. But it should be made clear somehow (e.g. in the function name) that the function should not be used in a production code path.

An alternative to copying the numbers from device to host for the printouts would be to launch a kernel that prints out the numbers.

Reply:
> But it should be made clear somehow (e.g. in the function name) that the function should not be used in a production code path.

I like this approach, for example adding a suffix `Sync` to names of the functions that call `alpaka::wait(queue)`.

Reply:
> I meant/implied that the current code copies from 1D (size 1) buffers.

Ah, OK.

Reply:
> > Some backends my _cause_ an extra synchronisation, but AFAIK not all of them do, and alpaka does not provide any such guarantee.
> 
> This is a good point. My earlier comment about "non-cached-allocated alpaka buffer synchronizing on its destructor" indeed was specific to NVIDIA and the behavior of `cudaFree()`.

Now that you mention it, I guess we should clarify the behaviour of the various types of buffers.

I will try to bring this up in one of the next alpaka meetings.